modules/dnsmasq_dhcp/autodetection.go

```diff
@@ -46,18 +60,128 @@ func (d *DnsmasqDHCP) findDHCPRanges() ([]string, error) {
 	return ranges, nil
 }
 
-func findConfigurations(confDir string) ([]string, error) {
+// findConfigs recursively finds and reads configuration files respecting
+// conf-file and conf-dir directives.
+// findConfigs is tolerant to IO errors and finds as maximum config
+// files as possible, therefore if an error occurrs during scanning process,
+// it will be just logged with warning severity
+func (d *DnsmasqDHCP) findConfigs(confPath string) []string {
+	config, err := os.Open(confPath)
+	if err != nil {
+		d.Warningf("error during configuration file %q reading: %v", confPath, err)
+		return nil
+	}
+
+	defer config.Close()
+
+	var (
+		includeFiles []string
+		includeDirs  []configDir
+	)
+
+	scanner := bufio.NewScanner(config)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if path, ok := getConfValue(line, "conf-file"); ok {
+			includeFiles = append(includeFiles, path)
+			continue
+		}
+
+		if path, ok := getConfValue(line, "conf-dir"); ok {
+			args := strings.Split(path, ",")
+
+			dir := configDir{
+				path: args[0],
+			}
+
+			for _, arg := range args[1:] {
+				arg := strings.TrimSpace(arg)
+				// dnsmasq treats suffixes with asterisk as "to include" and without
+				// asterisk as "to exclude"
+				if strings.HasPrefix(arg, "*") {
+					dir.includeSuffix = append(dir.includeSuffix, arg[1:])
+				} else {
+					dir.excludeSuffix = append(dir.excludeSuffix, arg)
+				}
+			}
+
+			includeDirs = append(includeDirs, dir)
+
+			continue
+		}
+	}
+
+	for _, dir := range includeDirs {
+		dirFiles, err := readConfDir(dir.path, dir.includeSuffix, dir.excludeSuffix)
+		if err != nil {
+			d.Warningf("error during configuration dir %q scanning: %v", dir.path, err)
+		}
+
+		includeFiles = append(includeFiles, dirFiles...)
+	}
+
+	for _, file := range includeFiles[:] {
+		files := d.findConfigs(file)
```

Comment:
what if `dnsmasq.conf` contains

> conf-file=/etc/dnsmasq.conf

?

why do we need `[:]` ?

Reply:
if config will have a recursion, we will run into limit of open files and os.Open() will fail eventually with error `too many open files`, but it's the same case as with dnsmasq itself.

We need `[:]` in order to avoid cycle on new arrived files found by d.findConfigs(), so we re-slice includeFiles and when we add new elements in `includeFiles` they will not appear in re-sliced slice. We could create a new slice and then merge them if you insist on avoiding that trick.

Reply:
> so we re-slice includeFiles and when we add new elements in includeFiles they will not appear in re-sliced slice

i think `range` already does it and `includeFiles` will point to another slice right after first append. Have you tested it?

Reply:
i may be wrong there ^^ :thinking: 	not sure

Reply:
```go
	vv := []int{1, 2, 3}

	println(vv)
	for _, v := range vv {
		vv = append(vv, v * 10)
		println(vv)
	}

	fmt.Println(vv)
```

```
[1 2 3 10 20 30]
[3/3]0xc000018540
[4/6]0xc00001a150  <- different slice
[5/6]0xc00001a150
[6/6]0xc00001a150
```

Reply:
ok, i think this is a bad example :sweat_smile: 

my point is  re-slicing is not needed in `range`

Reply:
ok, that is the internals of `for` loop

```
//   for_temp := range
//   len_temp := len(for_temp)
//   for index_temp = 0; index_temp < len_temp; index_temp++ {
//           value_temp = for_temp[index_temp]
//           index = index_temp
//           value = value_temp
//           original body
//   }
```

so, yes, we don't need re-slicing